### PR TITLE
rpm_key: support multiple siging keys

### PR DESCRIPTION
**Pull request by t-kranz**
##### SUMMARY

Fixes # #83394 

Took the changes from #75251 (for #50615)  - which implement checking for multiple key ids when importing keys with the  rpm_key module (see https://github.com/ansible/ansible/pull/75251#issue-944382323):

Make rpm_key module use lists for key ids since a single key file can contain multiple keys.

I also made the fingerprint option accept lists of fingerprints since they correspond with key ids. All fingerprints in a keyfile should be present when the fingerprints option is set for task success.

##### ISSUE TYPE

- Bugfix Pull Request

COMPONENT NAME

rpm_key

##### ADDITIONAL INFORMATION

Fixed the concerns from @sivel regarding # #75251:

- comparing the keys against the actual content of keyids instead of asserting the lengths are equal (https://github.com/ansible/ansible/pull/75251#discussion_r670746523)

- keeping the normalize_keyid function and creating a normalize_keyids funtion that calls the singular version for unit testing (https://github.com/ansible/ansible/pull/75251#discussion_r670747283)

Additional Info from #75251 

Example:

```paste below
- name: Add Google GPG keys to system
  rpm_key:
    key: https://dl.google.com/linux/linux_signing_key.pub
    fingerprint:
      # First key in keyfile
      - 4CCA 1EAF 950C EE4A B839 76DC A040 830F 7FAC 5991
      # Second key in keyfile
      - EB4C 1BFD 4F04 2F6D DDCC EC91 7721 F63B D38B 4796
```

The rpm_key module only checked for the presence of the first key in a file when importing keys from files. This meant that in cases where a single file contained multiple keys it was possible for later keys to not be imported if the first key in a key file was already installed. It could also mean that multiple keys could be installed even when only expecting a single key in a file.

### CODE REVIEW

**Comment by t-kranz** [comment]
I would like to remove the RHEL9 SHA1 deprecation from /test/integration/targets/rpm_key/tasks/rpm_key.yaml 

To be able to do that i guess a GPG key file would need to be added to https://ci-files.testing.ansible.com/test/integration/targets/

That file needs to contain keys with at least two different fingerprints and those keys have to comply with the SHA-1 deprecation in RHEL 9.

File: test/integration/targets/rpm_key/tasks/rpm_key.yaml [hunk 1]
```diff
@@ -187,6 +187,53 @@
...
+    dest: /tmp/fragments/RPM-GPG-KEY-PGDG-AARCH64-RHEL9
+
+- name: Issue 83493 - Assemble key from fragments
+  ansible.builtin.assemble:
+    src: /tmp/fragments
+    dest: /tmp/KEY_83493
+
+- name: Issue 83493 - Verify key fingerprints. This should fail due to multiple key signatures
+  rpm_key:
+    key: /tmp/KEY_83493
```
Comment (s-hertel) [inline test/integration/targets/rpm_key/tasks/rpm_key.yaml]: Thanks for working on this. You should be able to swap this out for https://ansible-ci-files.s3.amazonaws.com/test/integration/targets/rpm_key/linux_signing_key.pub now and remove the few tasks preceding this one.

**Review by s-hertel — APPROVED** [review]

File: changelogs/fragments/50615-rpm_key.yml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+minor_changes:
+  - >
+    rpm_key - Check for multiple signing keys in a single GPG key file.
```
Comment (webknjaz) [inline changelogs/fragments/50615-rpm_key.yml:3]: Could you rephrase this in non-imperative mood? The writing style is inconsistent with the following entry (and likely many other changelog fragments).
This is going to show up in the aggregated change log document on release. That document refers to the changes that have already happened since the previous release, so it's natural to describe things as if they are already merged. Imperative mood is a better fit for Git commit summary lines and TODO entries, but change logs have different style, purpose and audience.

File: lib/ansible/modules/rpm_key.py [hunk 3]
```diff
@@ -105,35 +106,37 @@ def __init__(self, module):
         state = module.params['state']
         key = module.params['key']
         fingerprint = module.params['fingerprint']
```
Comment (webknjaz) [inline lib/ansible/modules/rpm_key.py:108]: Since this is now a list, seeing code iterating over a variable called as a single item reads somewhat counterintuitive.
May I suggest renaming the variable to `fingerprints`? Even if you keep the public interface with the original name. This will make the code easier to reason about.

File: lib/ansible/modules/rpm_key.py [hunk 4]
```diff
@@ -105,35 +106,37 @@ def __init__(self, module):
...
 
         if state == 'present':
-            if self.is_key_imported(keyid):
+            if self.is_key_imported(keyids):
                 module.exit_json(changed=False)
             else:
                 if not keyfile:
                     self.module.fail_json(msg="When importing a key, a valid file must be given")
                 if fingerprint:
                     has_fingerprint = self.getfingerprint(keyfile)
```
Comment (webknjaz) [inline lib/ansible/modules/rpm_key.py:138]: This var should probably be called differently. Currently, `set(has_fingerprint)` looks as if a boolean is passed to the `set()` constructor. Perhaps, `computed_fingerprint`?

File: lib/ansible/modules/rpm_key.py [hunk 5]
```diff
@@ -105,35 +106,37 @@ def __init__(self, module):
...
-                    if fingerprint != has_fingerprint:
+                    if set(fingerprint) != set(has_fingerprint):
```
Comment (webknjaz) [inline lib/ansible/modules/rpm_key.py:139]: Should this be something along the lines of
```suggestion
                    if any(fgrprnt not in computed_fingerprint for fgrprnt in fingerprints):
```
?

Perhaps, this entire expression could be assigned to a variable, so its intended purpose could be labeled via a variable name?

File: lib/ansible/modules/rpm_key.py (hunk 3 above)
Comment (webknjaz) [reply lib/ansible/modules/rpm_key.py:108]: Or even `expected_fingerprints`?

File: lib/ansible/modules/rpm_key.py [hunk 6]
```diff
@@ -164,30 +167,42 @@ def fetch_key(self, url):
...
+        for keyid in keyids:
+            normalized_keyids.append(self.normalize_keyid(keyid))
+
+        return normalized_keyids
+
     def normalize_keyid(self, keyid):
         """Ensure a keyid doesn't have a leading 0x, has leading or trailing whitespace, and make sure is uppercase"""
         ret = keyid.strip().upper()
-        if ret.startswith('0x'):
+        if ret.startswith("0x"):
```
Comment (webknjaz) [inline lib/ansible/modules/rpm_key.py:180]: *Pro tip:* mixing formatting changes with functional ones is discouraged because it makes it harder for the reviewers to identify which places in the patch are useful payload. These should always be separated.

File: lib/ansible/modules/rpm_key.py [hunk 7]
```diff
@@ -164,30 +167,42 @@ def fetch_key(self, url):
...
+    def getkeyids(self, keyfile):
         stdout, stderr = self.execute_command([self.gpg, '--no-tty', '--batch', '--with-colons', '--fixed-list-mode', keyfile])
+        keyids = []
         for line in stdout.splitlines():
             line = line.strip()
             if line.startswith('pub:'):
-                return line.split(':')[4]
+                keyids.append(line.split(':')[4])
+
+        if keyids == []:
```
Comment (webknjaz) [inline lib/ansible/modules/rpm_key.py:195]: This is what people usually expect in Python:
```suggestion
        if not keyids:
```

File: lib/ansible/modules/rpm_key.py [hunk 8]
```diff
@@ -213,33 +231,45 @@ def execute_command(self, cmd):
...
+            imported_ids.add(imported_id)
+
+            # Also add "short" (rpm --queryformat %{version}) keyids to imported_ids
+            for keyid in keyids:
+                # if they are "short" id (self.is_keyid(keyid)) and substring of an imported_id
+                if self.is_keyid(keyid) and keyid in imported_id:
+                    imported_ids.add(keyid)
+
+        missing_keys = set(keyids).difference(imported_ids)
+        return len(missing_keys) == 0
```
Comment (webknjaz) [inline lib/ansible/modules/rpm_key.py:255]: this could be
```suggestion
        return not missing_keys
```

Comment (webknjaz) [inline lib/ansible/modules/rpm_key.py:255]: Though, I'd probably label the expression with what it intends to compute and return that:
```suggestion
        some_gpg_keys_missing = bool(set(keyids) - imported_ids)
        return some_gpg_keys_missing
```
(I think that “return whether the length of a set of missing keys is zero” might be more difficult to parse vs. “return whether some GPG keys are missing” — it's low level vs. high level)

File: test/integration/targets/rpm_key/tasks/rpm_key.yaml [hunk 9]
```diff
@@ -187,6 +187,33 @@
       - result is success
       - result is not changed
 
+- name: Issue 83493 - Verify key fingerprints. This should fail due to multiple key signatures
+  rpm_key:
+    key: https://ansible-ci-files.s3.amazonaws.com/test/integration/targets/rpm_key/linux_signing_key.pub
+    fingerprint: 1B09 0CE4 7562 66D7 1203 13F6 3BAC 309A 6EB4 25D5
```
Comment (deriamis) [inline test/integration/targets/rpm_key/tasks/rpm_key.yaml:193]: Correct me if I'm wrong, but isn't this a breaking change?

I think the current case is that users would only need to list one of the keys which appears in the import, and this test suggests that *all* keys would need to be listed for verification. I can imagine scenarios where users upload all their signing keys to a single file and then check for different keys in different modules. This change would break that usage if I have how it currently works right.
